api/src/middleware/authentication.ts

```diff
@@ -5,36 +5,47 @@ import * as response from 'lib/response';
 import * as userService from 'user/service';
 import * as authorizationService from 'authorization/service';
 
-const authentication = (optional = false, requiresName = true): middy.MiddlewareObj => {
+const authentication = (optional = false, requiresName = true, endpointSpecificKey?: string): middy.MiddlewareObj => {
     const before: middy.MiddlewareFn<I.APIGatewayProxyEventV2> = async (request): Promise<I.JSONResponse | void> => {
         try {
             let user: null | I.User = null;
 
             const apiKey = request.event.queryStringParameters?.apiKey;
-            const bearerToken = request.event.headers.Authorization;
 
-            if (apiKey) {
-                user = await userService.getByApiKey(apiKey);
-            } else if (bearerToken) {
-                user = authorizationService.validateJWT(bearerToken.split(' ')[1]);
-            }
-
-            if (!optional) {
-                // If there's no user account, and authentication is *not* optional, then the request is blocked.
-                if (!user) {
-                    return response.json(401, { message: 'Please enter either a valid apiKey or bearer token.' });
+            if (endpointSpecificKey) {
+                // The function only accepts one specific api key.)
+                if (apiKey !== endpointSpecificKey) {
+                    return response.json(401, { message: "Please provide a valid 'apiKey'." });
+                }
+            } else {
```

Comment:
Think you could do without the else block:

```
            if (endpointSpecificKey) {
                // The function only accepts one specific api key.)
                if (apiKey !== endpointSpecificKey) {
                    return response.json(401, { message: "Please provide a valid 'apiKey'." });
                } else {
                    return;
                }
            }

            const bearerToken = request.event.headers.Authorization;
            ...
```

From what I can figure out, you return 401/403 as you go along, then if the user makes it to the end, assign the user to the event object. In the case of endpoint specific key, user is null so you don't need to assign it? Unless you expect null specifically over undefined.

Reply:
Not sure I follow; the Object.assign won't get called if endpointSpecificKey is defined because of this else block. I think your way would have the same effect - either way, if endpointSpecificKey is defined it's going to ignore the rest from `const bearerToken ...` onwards.